diff --git a/setup.py b/setup.py
index b84171a..703646c 100755
--- a/setup.py
+++ b/setup.py
@@ -151,7 +151,6 @@ else:
     for include in os.environ.get("INCLUDEDIR", "").split(":"):
         _include_dirs.append(include)
 
-    _library_dirs += [sysconfig.get_config_var("LIBDIR"), '/opt/local/lib', '/usr/local/lib']
     if osx_check():
         _include_dirs += [
             '/opt/local/include/boost-1_35',
